File: app/src/main/java/ir/farshid_roohi/reminderpro/views/activities/BaseActivity.java
package ir.farshid_roohi.reminderpro.views.activities;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.content.pm.PackageManager;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import java.util.ArrayList;
import java.util.List;

import ir.farshid_roohi.reminderpro.ResourceApplication;

/**
 * Created by Farshid Roohi.
 * ReminderPro | Copyrights 2018.
 */
@SuppressLint("Registered")
public class BaseActivity extends AppCompatActivity {

    private static final int                         PERMISSION_REQUEST_ID = 901;
    private              OnPermissionRequestListener listener;

    private ResourceApplication resourceApplication;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.resourceApplication = ResourceApplication.getApplicationResource();

    }


    public <T extends ViewDataBinding> T setBindingView(@LayoutRes int layout) {
        T binding = DataBindingUtil.setContentView(this, layout);
        return binding;
    }


    public ResourceApplication getResourceApp() {
        return this.resourceApplication;
    }


    @TargetApi(Build.VERSION_CODES.M)
    public void requestPermission(String[] permissions, OnPermissionRequestListener listener) {


        this.listener = listener;

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            this.listener.onAllow("Current SDK Version lower than API 23");
            return;
        }


        List<String> permissionList = new ArrayList<>();


        for (String permission : permissions) {
            if (checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED) {
                this.listener.onAllow(permission);
                continue;
            }
            permissionList.add(permission);
        }

        if (permissionList.size() <= 0) {
            return;
        }
        requestPermissions(
                permissionList.toArray(new String[permissionList.size()])
                , PERMISSION_REQUEST_ID
        );


    }


    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {


        if (requestCode != PERMISSION_REQUEST_ID) {
            return;
        }

        for (int i = 0; i < permissions.length; i++) {

            if (grantResults[i] == PackageManager.PERMISSION_GRANTED) {
                this.listener.onAllow(permissions[i]);
                continue;
            }

            listener.onDenied(permissions[i]);

        }

    }

    public interface OnPermissionRequestListener {
        void onAllow(String permission);

        void onDenied(String permission);
    }
}

File: app/src/main/java/ir/roohi/farshid/reminderpro/views/activities/BaseActivity.kt
package ir.roohi.farshid.reminderpro.views.activities

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.SharedPreferences
import android.content.pm.PackageManager
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import android.os.Build
import android.os.Bundle
import android.preference.PreferenceManager
import android.util.Log
import android.view.Window
import androidx.annotation.ColorRes
import androidx.annotation.LayoutRes
import androidx.core.content.ContextCompat
import androidx.appcompat.app.AppCompatActivity
import android.view.WindowManager
import android.widget.Toast
import androidx.annotation.Nullable
import ir.roohi.farshid.reminderpro.ResourceApplication
import ir.roohi.farshid.reminderpro.listener.IEventBus
import ir.roohi.farshid.reminderpro.listener.OnPermissionRequestListener
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import java.util.*

/**
 * Created by Farshid Roohi.
 * ReminderPro | Copyrights 2018.
 */
@SuppressLint("Registered")
open class BaseActivity : AppCompatActivity(), IEventBus {

    private var listener: OnPermissionRequestListener? = null
    public var sharedPreferences: SharedPreferences? = null
    public var currentLanguage: String? = null

    companion object {

        private const val PERMISSION_REQUEST_ID = 901
    }

    var resourceApp: ResourceApplication? = null
        private set


    override fun onCreate(savedInstanceState: Bundle?) {
        this.sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this)
        this.currentLanguage = sharedPreferences!!.getString("LANGUAGE", "EN")
        setLocale(this.currentLanguage!!)
        super.onCreate(savedInstanceState)
        this.resourceApp = ResourceApplication.applicationResource

        if (!EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().register(this)
        }
    }

    fun setContentView(@LayoutRes layoutResID: Int, @ColorRes color: Int) {
        super.setContentView(layoutResID)
        setStatusBarColor(color)
    }

    fun hasFullScreen() {
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        this.window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN)
    }

    fun setStatusBarColor(@ColorRes color: Int) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            return
        }

        val window = this.window
        window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
        window.statusBarColor = contextCompatColor(color)
    }

    fun contextCompatColor(@ColorRes color: Int): Int {
        return ContextCompat.getColor(this, color)
    }


    fun <T : ViewDataBinding> setBindingView(@LayoutRes layout: Int): T {
        return DataBindingUtil.setContentView(this, layout)
    }

    fun showMsg(msg: String) {
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
    }

    @TargetApi(Build.VERSION_CODES.M)
    fun requestPermission(permissions: Array<String>, listener: OnPermissionRequestListener) {


        this.listener = listener

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            this.listener!!.onAllow("Current SDK Version lower than API 23")
            return
        }


        val permissionList = ArrayList<String>()


        for (permission in permissions) {
            if (checkPermissions(arrayOf(permission))) {
                this.listener!!.onAllow(permission)
                continue
            }
            permissionList.add(permission)
        }

        if (permissionList.size <= 0) {
            return
        }
        requestPermissions(permissionList.toTypedArray(), PERMISSION_REQUEST_ID)


    }


    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {


        if (requestCode != PERMISSION_REQUEST_ID) {
            return
        }

        for (i in permissions.indices) {

            if (grantResults[i] == PackageManager.PERMISSION_GRANTED) {
                this.listener!!.onAllow(permissions[i])
                continue
            }

            listener!!.onDenied(permissions[i])

        }

    }

    override fun onDestroy() {
        super.onDestroy()
        EventBus.getDefault().unregister(this)
    }

    @Subscribe
    override fun onChangeLanguage(str: String) {
        recreate()
    }


    open fun setLocale(lang: String) {
        val locale = Locale(lang)
        val res = resources
        val displayMatris = res.displayMetrics
        val configuration = res.configuration
        configuration.setLayoutDirection(locale)
        configuration.locale = locale
        res.updateConfiguration(configuration, displayMatris)

    }

    fun checkPermissions(permissions: Array<String>): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            return true
        }
        for (p in permissions) {
            if ((checkSelfPermission(p) != PackageManager.PERMISSION_GRANTED)) return false
        }
        return true
    }
}
